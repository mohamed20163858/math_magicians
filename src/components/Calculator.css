* {
  font-size: 25px;
  margin: 0;
}

.orange {
  background-color: #f5913e;
  border-top: 2px solid #c9c9c9;
}

.orange:hover {
  background-color: #c88e5f;
}

.orange:active {
  background-color: orange;
}

.gray {
  background-color: #e0e0e0;
  border-top: 2px solid #c9c9c9;
  border-right: 2px solid #c9c9c9;
}

.gray:hover {
  background-color: #b6b6b6;
}

.gray:active {
  background-color: white;
}

#root {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
}

.calculator {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  width: 50%;
  height: 50vh;
}

p {
  grid-column: 1 / 5;
  color: white;
  background-color: #858694;
  display: flex;
  justify-content: flex-end;
  padding-right: 10px;
  align-items: center;
  margin: 0;
}

button {
  border: none;
}

button:nth-of-type(17) {
  grid-column: 1 / 3;
}
