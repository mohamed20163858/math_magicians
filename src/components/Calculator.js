import React, { useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../logic/calculate';
import './Calculator.css';

const Button = (props) => {
  const { name, myClass, recievedInput } = props;
  const processRecievedInput = (e) => {
    recievedInput(e.target.textContent);
  };

  return (
    <button type="submit" className={myClass} onClick={processRecievedInput}>{name}</button>
  );
};
Button.propTypes = {
  name: PropTypes.string.isRequired,
  myClass: PropTypes.string.isRequired,
  recievedInput: PropTypes.string.isRequired,
};
const Paragraph = (props) => {
  let result = '';
  const { total, operation, next } = props;
  if (total) {
    result += `${total} `;
  }

  if (operation) {
    result += `${operation} `;
  }
  if (next) {
    result += `${next} `;
  }
  return (
    <p>{result}</p>
  );
};
Paragraph.propTypes = {
  total: PropTypes.string.isRequired,
  operation: PropTypes.string.isRequired,
  next: PropTypes.string.isRequired,
};
const Calculator = () => {
  const initialS = {
    total: null,
    next: null,
    operation: null,
  };
  const [s, setS] = useState(initialS);
  const passInput = (buttonName) => {
    setS(calculate(s, buttonName));
  };

  const symbols = ['AC', '+/-', '%', 'รท', '7', '8', '9', 'x', '4', '5', '6', '-', '1', '2', '3', '+', '0', '.', '='];
  const buttons = [];
  for (let i = 0; i < symbols.length; i += 1) {
    if ((i + 1) % 4 === 0 || (i + 1) === symbols.length) {
      buttons.push(<Button name={symbols[i]} recievedInput={passInput} myClass="orange" />);
    } else {
      buttons.push(<Button name={symbols[i]} recievedInput={passInput} myClass="gray" />);
    }
  }
  return (
    <div id="body">
      <h3>Let&apos;s do some math!</h3>
      <div className="calculator">
        <Paragraph operation={s.operation} next={s.next} total={s.total} />
        {buttons}
      </div>
    </div>
  );
};
export default Calculator;
