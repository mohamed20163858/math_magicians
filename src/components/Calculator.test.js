import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import Calculator from './Calculator';

it('renders correctly', () => {
  const tree = renderer
    .create(<Calculator />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('simulate adding two numbers', () => {
  // ARRANGE
  render(<Calculator />);

  // ACT
  userEvent.click(screen.getByText('9'));
  userEvent.click(screen.getByText('+'));
  userEvent.click(screen.getByText('5'));
  userEvent.click(screen.getByText('='));
  const result = screen.getByText('14');

  // ASSERT
  expect(result).toBeInTheDocument();
});
