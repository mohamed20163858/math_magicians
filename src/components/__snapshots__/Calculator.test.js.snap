// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="body"
>
  <h3>
    Let's do some math!
  </h3>
  <div
    className="calculator"
  >
    <p />
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      AC
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      +/-
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      %
    </button>
    <button
      className="orange"
      onClick={[Function]}
      type="submit"
    >
      รท
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      7
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      8
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      9
    </button>
    <button
      className="orange"
      onClick={[Function]}
      type="submit"
    >
      x
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      4
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      5
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      6
    </button>
    <button
      className="orange"
      onClick={[Function]}
      type="submit"
    >
      -
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      1
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      2
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      3
    </button>
    <button
      className="orange"
      onClick={[Function]}
      type="submit"
    >
      +
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      0
    </button>
    <button
      className="gray"
      onClick={[Function]}
      type="submit"
    >
      .
    </button>
    <button
      className="orange"
      onClick={[Function]}
      type="submit"
    >
      =
    </button>
  </div>
</div>
`;
